apps/src/AzureTextToSpeech.js

```diff
@@ -0,0 +1,252 @@
+   */
+  static getSingleton() {
+    if (!singleton) {
+      singleton = new AzureTextToSpeech();
+    }
+    return singleton;
+  }
+
+  constructor() {
+    this.playing = false;
+    this.queue_ = [];
+    this.cachedSounds_ = {};
+    this.playbackOptions_ = {
+      volume: 1.0,
+      loop: false,
+      forceHTML5: false,
+      allowHTML5Mobile: true,
+      onEnded: this.onSoundComplete_
+    };
+  }
+
+  /**
+   *
+   * @param {Promise<SoundResponse>} soundPromise A promise that returns a SoundResponse when resolved.
+   * @param {function(ArrayBuffer, Object)} play Function that plays sound bytes. Object parameter accepts
+   * playback configuration options.
+   */
+  enqueueAndPlay = (soundPromise, play) => {
+    this.enqueue_(soundPromise);
+    this.asyncPlayFromQueue_(play);
+  };
+
+  /**
+   * Returns a promise representing a TTS sound that can be enqueued and played. Utilizes a sound cache --
+   * will check for a cache hit to avoid duplicate network requests, and caches network responses for re-use.
+   * @param {Object} opts
+   * @param {string} opts.text
+   * @param {string} opts.gender
+   * @param {string} opts.languageCode
+   * @param {string} opts.url URL to request sound from.
+   * @param {string} opts.ssml SSML in request body.
+   * @param {string} opts.token Authentication token from Azure.
+   * @param {function(Array<string>)} opts.onProfanityFound Called if the given text contains profanity.
+   * @returns {Promise<SoundResponse>} A promise that returns a SoundResponse when resolved.
+   */
+  createSoundPromise = opts => {
+    const {
+      text,
+      gender,
+      languageCode,
+      url,
+      ssml,
+      token,
+      onProfanityFound
+    } = opts;
+    const cachedSound = this.getCachedSound_(languageCode, gender, text);
+    const wrappedSetCachedSound = soundResponse => {
+      this.setCachedSound_(languageCode, gender, text, soundResponse);
+    };
+    const wrappedCreateSoundResponse = this.createSoundResponse_;
+
+    // If we have the sound already, resolve immediately.
+    if (cachedSound) {
+      const {bytes, profaneWords} = cachedSound;
+
+      return new Promise(resolve => {
+        if (profaneWords && profaneWords.length > 0) {
+          onProfanityFound(profaneWords);
+          resolve(wrappedCreateSoundResponse({profaneWords}));
+        } else {
+          resolve(wrappedCreateSoundResponse({bytes}));
+        }
+      });
+    }
+
+    // Otherwise, check the text for profanity and request the TTS sound.
+    return new Promise(async resolve => {
+      const profaneWords = await findProfanity(
+        text,
+        languageCode,
+        true /* shouldCache */
+      );
+
+      if (profaneWords && profaneWords.length > 0) {
+        onProfanityFound(profaneWords);
+        const soundResponse = wrappedCreateSoundResponse({profaneWords});
+        wrappedSetCachedSound(soundResponse);
+        resolve(soundResponse);
+        return;
+      }
```

Comment:
It looks like we still do a bunch of "normal" processing when profanity is detected.  Could you add some description of what happens when profanity is detected, maybe at the top of the file where `SoundResponse` is described?

Reply:
Ah, I see that we actually do prevent any further processing if we detect profane words.  Thanks for the added commentary!